diff --git a/www/wwwsh b/www/wwwsh
index 7f56332..a830a95 100755
--- a/www/wwwsh
+++ b/www/wwwsh
@@ -24,7 +24,8 @@ fi
 if [ ${CONTENT_LENGTH:-0} -gt 0 ]; then
     POST_FILE=/tmp/post.$$
     export POST_FILE
-    dd of=$POST_FILE cbs=$CONTENT_LENGTH conv=block count=1 2>/dev/null
+#    dd of=$POST_FILE cbs=$CONTENT_LENGTH conv=block count=1 2>/dev/null
+    dd of=$POST_FILE count=$CONTENT_LENGTH bs=1 2>/dev/null
     $haserl $file < $POST_FILE
     /bin/rm -f $POST_FILE
 else
