--- feeds/luci/luci/Makefile	2014-10-26 10:58:06.665834346 +0000
+++ feeds/luci/luci/Makefile	2014-10-26 11:02:30.705537902 +0000
@@ -215,7 +215,7 @@
 $(eval $(call protocol,ipv6,Support for DHCPv6/6in4/6to4/6rd/DS-Lite))
 $(eval $(call protocol,3g,Support for 3G,+PACKAGE_luci-proto-3g:comgt))
 $(eval $(call protocol,relay,Support for relayd pseudo bridges,+PACKAGE_luci-proto-relay:relayd))
-
+$(eval $(call protocol,ncm,Support for NCM protocol:ncm))
 
 ### Community Packages ###
 define Package/luci-mod-freifunk-community
@@ -338,6 +338,9 @@
 $(eval $(call application,meshwizard, Shellscript based wizard to setup mesh networks,\
         +PACKAGE_luci-app-freifunk-meshwizard:meshwizard))
 
+$(eval $(call application,ncm-status, Diagnostics for NCM dongle,\
+	+PACKAGE_luci-app-ncm-status:luci-proto-ncm))
+
 $(eval $(call application,olsr,OLSR configuration and status module,\
 	+olsrd-mod-txtinfo))
 
